### PR TITLE
Add CA DMV behind-the-wheel appointment example.

### PR DESCRIPTION
**Pull request by a1ph**
The script allows you to pick up a certain date for your DMV
behind-the-wheel test appointment.

### CODE REVIEW

File: examples/ca-dmv-appointment.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2017 Google Inc., PhantomJS Authors All rights reserved.
```
Comment (paulirish) [inline examples/ca-dmv-appointment.js:2]: afaik just google on this header

File: examples/ca-dmv-appointment.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  }
+}
+
+async function fillForm(page, location) {
+  await page.navigate('https://www.dmv.ca.gov/wasapp/foa/clear.do?goTo=driveTest&localeName=en');
+  await sleep(500);
+  await page.waitFor('.form-horizontal');
+  formData.location = location;
+  await page.evaluate(formData => {
+    const offices = $('select#officeId option');
```
Comment (paulirish) [inline examples/ca-dmv-appointment.js:72]: is the `$` coming from our commandline API, or is it their copy of jQuery?

File: examples/ca-dmv-appointment.js (hunk 1 above)
Comment (aslushnikov) [inline examples/ca-dmv-appointment.js:2]: let's have just google's license here

File: examples/ca-dmv-appointment.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  await page.screenshot({path: 'dmv-filed-form.png'});
+
+  // Sumbitting the form
+  await page.evaluate(() => $('.form-horizontal input[type="submit"]').click());
+
+  try {
+    await page.waitFor('#ApptForm');
+  } catch (e) {
+    await page.screenshot({path: 'dmv-form-response.png'});
+    console.log('Form data seems to be wrong. Please check the form response screenshot.');
```
Comment (paulirish) [inline examples/ca-dmv-appointment.js:105]: love taking a screenshot in the error state. nice

File: examples/ca-dmv-appointment.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    await page.screenshot({path: 'dmv.png'});
+    return false;
+  }
+
+  // Almost there. Confirm the appointment.
+  await page.evaluate(() => $('form.form-horizontal[id="ApptForm"] input[type="submit"]').click());
+  await sleep(500);
+  await page.waitFor('#ApptForm');
+  await page.screenshot({path: 'dmv-confirmation.png'});
+  await page.evaluate(() => $('form.form-horizontal[id="ApptForm"] input[value="Confirm"]').click());
```
Comment (aslushnikov) [inline examples/ca-dmv-appointment.js:128]: Should page.click work here? 
```js
page.click('form.form-horizontal[id="ApptForm"] input[value="Confirm"]')
```

File: examples/ca-dmv-appointment.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  await sleep(2000);
+  await page.screenshot({path: 'dmv-confirmed.png'});
+  return true;
+}
+
+async function sleep(timeout) {
```
Comment (aslushnikov) [inline examples/ca-dmv-appointment.js:134]: can we kill sleeps altogether? they should not be needed

File: examples/ca-dmv-appointment.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    document.getElementById('dl_number').value = formData.licenseId;
+    document.getElementsByName('birthMonth')[0].value = formData.birthDate.split('-')[0];
+    document.getElementsByName('birthDay')[0].value = formData.birthDate.split('-')[1];
+    document.getElementsByName('birthYear')[0].value = formData.birthDate.split('-')[2];
+    document.getElementsByName('telArea')[0].value = formData.phone.split('-')[0];
+    document.getElementsByName('telPrefix')[0].value = formData.phone.split('-')[1];
+    document.getElementsByName('telSuffix')[0].value = formData.phone.split('-')[2];
+  }, formData);
+
+  await page.screenshot({path: 'dmv-filed-form.png'});
```
Comment (aslushnikov) [inline examples/ca-dmv-appointment.js:96]: Looks like you're interested in a full-page screenshot everywhere.. shall we make it a default? cc @ebidel @paulirish

File: examples/ca-dmv-appointment.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  'los gatos',
+];
+const timeout = 30; // seconds between retries.
+
+// ===== Edit fields above this line
+
+const Browser = require('../lib/Browser');
+const browser = new Browser();
+
+browser.newPage().then(async page => {
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:46]: to be consistent with the other examples:

```javascript
(async() => {
  const page = await browser.newPage();
  await page.setViewport({width: 600, height: 1200});
  await iterate(page);
  browser.close();
})();
```

**Review by aslushnikov — COMMENTED** [review]
Success story! 😏 Would you mind changing a few bits here and there to make it a nice api demo?

File: examples/ca-dmv-appointment.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    document.getElementsByName('birthDay')[0].value = formData.birthDate.split('-')[1];
+    document.getElementsByName('birthYear')[0].value = formData.birthDate.split('-')[2];
+    document.getElementsByName('telArea')[0].value = formData.phone.split('-')[0];
+    document.getElementsByName('telPrefix')[0].value = formData.phone.split('-')[1];
+    document.getElementsByName('telSuffix')[0].value = formData.phone.split('-')[2];
+  }, formData);
+
+  await page.screenshot({path: 'dmv-filed-form.png'});
+
+  // Sumbitting the form
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:98]: // Submit the form

File: examples/ca-dmv-appointment.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  try {
+    await page.waitFor('#ApptForm');
+  } catch (e) {
+    await page.screenshot({path: 'dmv-form-response.png'});
+    console.log('Form data seems to be wrong. Please check the form response screenshot.');
+    return true;
+  }
+
+  // Response is here. Check if the date works for us.
+  const officeString = await page.evaluate(() => $('td[data-title="Office"] p').text());
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:110]: Is jquery on their page? That might be too confusing :) Probably should stick with vanilla:

`document.querySelector('td[data-title="Office"] p').textContent`
`document.querySelector('td[data-title="Appointment"] strong').textContent`

File: examples/ca-dmv-appointment.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        if (res)
+          return;
+      } catch (e) {}
+      console.log(`\nSleeping for ${timeout} sec...`);
+      await sleep(timeout * 1e3);
+    }
+  }
+}
+
+async function fillForm(page, location) {
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:66]: it would be nice to have jsdocs on these functions

File: examples/ca-dmv-appointment.js (hunk 5 above)
Comment (ebidel) [inline examples/ca-dmv-appointment.js:134]: dont think you need the `async`

File: examples/ca-dmv-appointment.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  await page.screenshot({path: 'dmv-confirmation.png'});
+  await page.evaluate(() => $('form.form-horizontal[id="ApptForm"] input[value="Confirm"]').click());
+  await sleep(2000);
+  await page.screenshot({path: 'dmv-confirmed.png'});
+  return true;
+}
+
+async function sleep(timeout) {
+  return new Promise(r => setTimeout(r, timeout));
+}
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:136]: newline

File: examples/ca-dmv-appointment.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  'san mateo',
+  'san jose',
+  'redwood city',
+  'los gatos',
+];
+const timeout = 30; // seconds between retries.
+
+// ===== Edit fields above this line
+
+const Browser = require('../lib/Browser');
```
Comment (ebidel) [inline examples/ca-dmv-appointment.js:43]: Can we stick all requires at the top?

Comment (ebidel) [reply examples/ca-dmv-appointment.js:43]: Oh my, we've got 3 flavors of this require in the codebase so far:
```
const {Browser} = require('puppeteer');
const {Browser} = require('..');
const Browser = require('../lib/Browser');
```

Suggest sticking with `const {Browser} = require('puppeteer');` in all the examples. That'll make it easier to cut and paste from the examples.

**Review by ebidel — CHANGES_REQUESTED** [review]

**Comment by aslushnikov** [comment]
We feel uncomfortable having this in – let's not add it.

**Comment by Garbee** [comment]
I think this is a decent example that shows you can pre-populate data in some way (the `FormData`) as well as capture screenshots on error for the developer to view what went wrong. Overall, quite useful for people to have a reference point to know it is possible.

What is the source of the uncomfort with including this? Is there anything that can be done to the example to make it acceptable (beyond all of the existing review on code style.)

**Pull request closed without merging**